# Venture Funding with Deep Learning

You work as a risk management associate at Alphabet Soup, a venture capital firm. Alphabet Soup’s business team receives many funding applications from startups every day. This team has asked you to help them create a model that predicts whether applicants will be successful if funded by Alphabet Soup.

The business team has given you a CSV containing more than 34,000 organizations that have received funding from Alphabet Soup over the years. With your knowledge of machine learning and neural networks, you decide to use the features in the provided dataset to create a binary classifier model that will predict whether an applicant will become a successful business. The CSV file contains a variety of information about these businesses, including whether or not they ultimately became successful.

## Instructions:

The steps for this challenge are broken out into the following sections:

* Prepare the data for use on a neural network model.

* Compile and evaluate a binary classification model using a neural network.

* Optimize the neural network model.

### Prepare the Data for Use on a Neural Network Model 

Using your knowledge of Pandas and scikit-learn’s `StandardScaler()`, preprocess the dataset so that you can use it to compile and evaluate the neural network model later.

Open the starter code file, and complete the following data preparation steps:

1. Read the `applicants_data.csv` file into a Pandas DataFrame. Review the DataFrame, looking for categorical variables that will need to be encoded, as well as columns that could eventually define your features and target variables.   

2. Drop the “EIN” (Employer Identification Number) and “NAME” columns from the DataFrame, because they are not relevant to the binary classification model.
 
3. Encode the dataset’s categorical variables using `OneHotEncoder`, and then place the encoded variables into a new DataFrame.

4. Add the original DataFrame’s numerical variables to the DataFrame containing the encoded variables.

> **Note** To complete this step, you will employ the Pandas `concat()` function that was introduced earlier in this course. 

5. Using the preprocessed data, create the features (`X`) and target (`y`) datasets. The target dataset should be defined by the preprocessed DataFrame column “IS_SUCCESSFUL”. The remaining columns should define the features dataset. 

6. Split the features and target sets into training and testing datasets.

7. Use scikit-learn's `StandardScaler` to scale the features data.

### Compile and Evaluate a Binary Classification Model Using a Neural Network

Use your knowledge of TensorFlow to design a binary classification deep neural network model. This model should use the dataset’s features to predict whether an Alphabet Soup&ndash;funded startup will be successful based on the features in the dataset. Consider the number of inputs before determining the number of layers that your model will contain or the number of neurons on each layer. Then, compile and fit your model. Finally, evaluate your binary classification model to calculate the model’s loss and accuracy. 
 
To do so, complete the following steps:

1. Create a deep neural network by assigning the number of input features, the number of layers, and the number of neurons on each layer using Tensorflow’s Keras.

> **Hint** You can start with a two-layer deep neural network model that uses the `relu` activation function for both layers.

2. Compile and fit the model using the `binary_crossentropy` loss function, the `adam` optimizer, and the `accuracy` evaluation metric.

> **Hint** When fitting the model, start with a small number of epochs, such as 20, 50, or 100.

3. Evaluate the model using the test data to determine the model’s loss and accuracy.

4. Save and export your model to an HDF5 file, and name the file `AlphabetSoup.h5`. 

### Optimize the Neural Network Model

Using your knowledge of TensorFlow and Keras, optimize your model to improve the model's accuracy. Even if you do not successfully achieve a better accuracy, you'll need to demonstrate at least two attempts to optimize the model. You can include these attempts in your existing notebook. Or, you can make copies of the starter notebook in the same folder, rename them, and code each model optimization in a new notebook. 

> **Note** You will not lose points if your model does not achieve a high accuracy, as long as you make at least two attempts to optimize the model.

To do so, complete the following steps:

1. Define at least three new deep neural network models (the original plus 2 optimization attempts). With each, try to improve on your first model’s predictive accuracy.

> **Rewind** Recall that perfect accuracy has a value of 1, so accuracy improves as its value moves closer to 1. To optimize your model for a predictive accuracy as close to 1 as possible, you can use any or all of the following techniques:
>
> * Adjust the input data by dropping different features columns to ensure that no variables or outliers confuse the model.
>
> * Add more neurons (nodes) to a hidden layer.
>
> * Add more hidden layers.
>
> * Use different activation functions for the hidden layers.
>
> * Add to or reduce the number of epochs in the training regimen.

2. After finishing your models, display the accuracy scores achieved by each model, and compare the results.

3. Save each of your models as an HDF5 file.


In [44]:
# Imports
import pandas as pd
from pathlib import Path
import tensorflow as tf
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler,OneHotEncoder

---

## Prepare the data to be used on a neural network model

### Step 1: Read the `applicants_data.csv` file into a Pandas DataFrame. Review the DataFrame, looking for categorical variables that will need to be encoded, as well as columns that could eventually define your features and target variables.  


In [45]:
# Read the applicants_data.csv file from the Resources folder into a Pandas DataFrame
applicant_data_df = pd.read_csv(Path('Resources/applicants_data.csv'))

# Review the DataFrame
applicant_data_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [46]:
# Review the data types associated with the columns
applicant_data_df.dtypes


EIN                        int64
NAME                      object
APPLICATION_TYPE          object
AFFILIATION               object
CLASSIFICATION            object
USE_CASE                  object
ORGANIZATION              object
STATUS                     int64
INCOME_AMT                object
SPECIAL_CONSIDERATIONS    object
ASK_AMT                    int64
IS_SUCCESSFUL              int64
dtype: object

### Step 2: Drop the “EIN” (Employer Identification Number) and “NAME” columns from the DataFrame, because they are not relevant to the binary classification model.

In [47]:
# Drop the 'EIN' and 'NAME' columns from the DataFrame
applicant_data_df = applicant_data_df.drop(columns =['EIN' , 'NAME'], axis = 1)

# Review the DataFrame



In [48]:
applicant_data_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

### Step 3: Encode the dataset’s categorical variables using `OneHotEncoder`, and then place the encoded variables into a new DataFrame.

In [49]:
# Create a list of categorical variables 
categorical_variables = list(applicant_data_df.dtypes[applicant_data_df.dtypes == 'object'].index)

# Display the categorical variables list
categorical_variables

['APPLICATION_TYPE',
 'AFFILIATION',
 'CLASSIFICATION',
 'USE_CASE',
 'ORGANIZATION',
 'INCOME_AMT',
 'SPECIAL_CONSIDERATIONS']

In [50]:
# Create a OneHotEncoder instance
enc = OneHotEncoder(sparse=False)

In [51]:
# Encode the categorcal variables using OneHotEncoder
encoded_data = enc.fit_transform(applicant_data_df[categorical_variables])

In [52]:
# Create a DataFrame with the encoded variables
encoded_df = pd.DataFrame(encoded_data,
                          columns = enc.get_feature_names(categorical_variables))

# Review the DataFrame
encoded_df.head()


C:\Users\domto\anaconda3\envs\dev\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  \
0                   1.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T14  APPLICATION_TYPE_T15  APPLICATION_TYPE_T17  \
0                   0.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  APPLICATION_TYPE_T25  \
0                   0.0                  0.0                   0.0   
1                   0.0                  0.0                   0.0   
2                   0.0                  0.0                   0.0   
3                   0.0                  0.0                   0.0   
4                   0.0                  0.0                   0.0   

   APPLICATION_TYPE_T29  ...  INCOME_AMT_1-9999  INCOME_AMT_10000-24999  \
0                   0.0  ...                0.0                     0.0   
1                   0.0  ...                1.0                     0.0   
2                   0.0  ...                0.0                     0.0   
3                   0.0  ...                0.0                     1.0   
4                   0.0  ...                0.0                     0.0   

   INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  \
0                       0.0                 0.0               0.0   
1                       0.0                 0.0               0.0   
2                       0.0                 0.0               0.0   
3                       0.0                 0.0               0.0   
4                       1.0                 0.0               0.0   

   INCOME_AMT_25000-99999  INCOME_AMT_50M+  INCOME_AMT_5M-10M  \
0                     0.0              0.0                0.0   
1                     0.0              0.0                0.0   
2                     0.0              0.0                0.0   
3                     0.0              0.0                0.0   
4                     0.0              0.0                0.0   

   SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  
0                       1.0                       0.0  
1                       1.0                       0.0  
2                       1.0                       0.0  
3                       1.0                       0.0  
4                       1.0                       0.0  

[5 rows x 114 columns]

### Step 4: Add the original DataFrame’s numerical variables to the DataFrame containing the encoded variables.

> **Note** To complete this step, you will employ the Pandas `concat()` function that was introduced earlier in this course. 

In [53]:
# Add the numerical variables from the original DataFrame to the one-hot encoding DataFrame

numerical_variable_df = applicant_data_df.drop(columns = categorical_variables)

attrition_df = pd.concat([
    numerical_variable_df,
    encoded_df], axis=1
)

# Review the Dataframe
attrition_df.head()


STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  \
0       1     5000              1                   1.0                   0.0   
1       1   108590              1                   0.0                   0.0   
2       1     5000              0                   0.0                   0.0   
3       1     6692              1                   0.0                   0.0   
4       1   142590              1                   0.0                   0.0   

   APPLICATION_TYPE_T13  APPLICATION_TYPE_T14  APPLICATION_TYPE_T15  \
0                   0.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T17  APPLICATION_TYPE_T19  ...  INCOME_AMT_1-9999  \
0                   0.0                   0.0  ...                0.0   
1                   0.0                   0.0  ...                1.0   
2                   0.0                   0.0  ...                0.0   
3                   0.0                   0.0  ...                0.0   
4                   0.0                   0.0  ...                0.0   

   INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  \
0                     0.0                       0.0                 0.0   
1                     0.0                       0.0                 0.0   
2                     0.0                       0.0                 0.0   
3                     1.0                       0.0                 0.0   
4                     0.0                       1.0                 0.0   

   INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  INCOME_AMT_50M+  \
0               0.0                     0.0              0.0   
1               0.0                     0.0              0.0   
2               0.0                     0.0              0.0   
3               0.0                     0.0              0.0   
4               0.0                     0.0              0.0   

   INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  
0                0.0                       1.0                       0.0  
1                0.0                       1.0                       0.0  
2                0.0                       1.0                       0.0  
3                0.0                       1.0                       0.0  
4                0.0                       1.0                       0.0  

[5 rows x 117 columns]

### Step 5: Using the preprocessed data, create the features (`X`) and target (`y`) datasets. The target dataset should be defined by the preprocessed DataFrame column “IS_SUCCESSFUL”. The remaining columns should define the features dataset. 



In [54]:
# Define the target set y using the IS_SUCCESSFUL column
y = attrition_df['IS_SUCCESSFUL']

# Display a sample of y
y[:5]


0    1
1    1
2    0
3    1
4    1
Name: IS_SUCCESSFUL, dtype: int64

In [55]:
# Define features set X by selecting all columns but IS_SUCCESSFUL
X = attrition_df.drop(columns = ['IS_SUCCESSFUL'])

# Review the features DataFrame
X.head()


STATUS  ASK_AMT  APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  \
0       1     5000                   1.0                   0.0   
1       1   108590                   0.0                   0.0   
2       1     5000                   0.0                   0.0   
3       1     6692                   0.0                   0.0   
4       1   142590                   0.0                   0.0   

   APPLICATION_TYPE_T13  APPLICATION_TYPE_T14  APPLICATION_TYPE_T15  \
0                   0.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T17  APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  ...  \
0                   0.0                   0.0                  0.0  ...   
1                   0.0                   0.0                  0.0  ...   
2                   0.0                   0.0                  0.0  ...   
3                   0.0                   0.0                  0.0  ...   
4                   0.0                   0.0                  0.0  ...   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                0.0                     0.0                       0.0   
1                1.0                     0.0                       0.0   
2                0.0                     0.0                       0.0   
3                0.0                     1.0                       0.0   
4                0.0                     0.0                       1.0   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                 0.0               0.0                     0.0   
1                 0.0               0.0                     0.0   
2                 0.0               0.0                     0.0   
3                 0.0               0.0                     0.0   
4                 0.0               0.0                     0.0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0              0.0                0.0                       1.0   
1              0.0                0.0                       1.0   
2              0.0                0.0                       1.0   
3              0.0                0.0                       1.0   
4              0.0                0.0                       1.0   

   SPECIAL_CONSIDERATIONS_Y  
0                       0.0  
1                       0.0  
2                       0.0  
3                       0.0  
4                       0.0  

[5 rows x 116 columns]

### Step 6: Split the features and target sets into training and testing datasets.


In [56]:
# Split the preprocessed data into a training and testing dataset
# Assign the function a random_state equal to 1
X_train, X_test, y_train, y_test = train_test_split(X,y, random_state=1)


### Step 7: Use scikit-learn's `StandardScaler` to scale the features data.

In [57]:
# Create a StandardScaler instance
scaler = StandardScaler()

# Fit the scaler to the features training dataset
X_scaler = scaler.fit(X_train)

# Fit the scaler to the features training dataset
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)


---

## Compile and Evaluate a Binary Classification Model Using a Neural Network

### Step 1: Create a deep neural network by assigning the number of input features, the number of layers, and the number of neurons on each layer using Tensorflow’s Keras.

> **Hint** You can start with a two-layer deep neural network model that uses the `relu` activation function for both layers.


In [132]:
# Define the the number of inputs (features) to the model
number_input_features = 116

# Review the number of features
number_input_features


116

In [143]:
# Define the number of neurons in the output layer
number_output_neurons = 1

In [144]:
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1 = 22

# Review the number hidden nodes in the first layer
hidden_nodes_layer1


22

In [145]:
# Define the number of hidden nodes for the second hidden layer
hidden_nodes_layer2 =  11

# Review the number hidden nodes in the second layer
hidden_nodes_layer2


11

In [146]:
# Create the Sequential model instance
nn = Sequential()

In [147]:
# Add the first hidden layer
nn.add(Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation = 'relu'))


In [148]:
# Add the second hidden layer
nn.add(Dense(units=hidden_nodes_layer2, activation='relu'))


In [149]:
# Add the output layer to the model specifying the number of output neurons and activation function
nn.add(Dense(units=number_output_neurons,activation='sigmoid'))


In [150]:
# Display the Sequential model summary
nn.summary()


Model: "sequential_10"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_29 (Dense)            (None, 22)                2574      
                                                                 
 dense_30 (Dense)            (None, 11)                253       
                                                                 
 dense_31 (Dense)            (None, 1)                 12        
                                                                 
Total params: 2,839
Trainable params: 2,839
Non-trainable params: 0
_________________________________________________________________


### Step 2: Compile and fit the model using the `binary_crossentropy` loss function, the `adam` optimizer, and the `accuracy` evaluation metric.


In [151]:
# Compile the Sequential model
nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])


In [152]:
# Fit the model using 50 epochs and the training data
model = nn.fit(X_train_scaled,y_train,epochs=50)

Epoch 1/50
804/804 [==============================] - 1s 599us/step - loss: 0.5815 - accuracy: 0.7124
Epoch 2/50
804/804 [==============================] - 0s 600us/step - loss: 0.5550 - accuracy: 0.7261
Epoch 3/50
804/804 [==============================] - 0s 584us/step - loss: 0.5512 - accuracy: 0.7285
Epoch 4/50
804/804 [==============================] - 0s 582us/step - loss: 0.5487 - accuracy: 0.7299
Epoch 5/50
804/804 [==============================] - 0s 585us/step - loss: 0.5475 - accuracy: 0.7315
Epoch 6/50
804/804 [==============================] - 0s 604us/step - loss: 0.5459 - accuracy: 0.7308
Epoch 7/50
804/804 [==============================] - 0s 584us/step - loss: 0.5454 - accuracy: 0.7325
Epoch 8/50
804/804 [==============================] - 0s 588us/step - loss: 0.5448 - accuracy: 0.7325
Epoch 9/50
804/804 [==============================] - 0s 584us/step - loss: 0.5440 - accuracy: 0.7331
Epoch 10/50
804/804 [==============================] - 0s 588us/step - loss: 0.543

### Step 3: Evaluate the model using the test data to determine the model’s loss and accuracy.


In [199]:
# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled, y_test, verbose=2)

# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5523 - accuracy: 0.7312 - 108ms/epoch - 403us/step
Loss: 0.5523036122322083, Accuracy: 0.731195330619812


### Step 4: Save and export your model to an HDF5 file, and name the file `AlphabetSoup.h5`. 


In [200]:
# Set the model's file path
file_path = Path('Resources/AlphabetSoup.h5')

# Export your model to a HDF5 file
nn.save(file_path)


---

## Optimize the neural network model


### Step 1: Define at least three new deep neural network models (resulting in the original plus 3 optimization attempts). With each, try to improve on your first model’s predictive accuracy.

> **Rewind** Recall that perfect accuracy has a value of 1, so accuracy improves as its value moves closer to 1. To optimize your model for a predictive accuracy as close to 1 as possible, you can use any or all of the following techniques:
>
> * Adjust the input data by dropping different features columns to ensure that no variables or outliers confuse the model.
>
> * Add more neurons (nodes) to a hidden layer.
>
> * Add more hidden layers.
>
> * Use different activation functions for the hidden layers.
>
> * Add to or reduce the number of epochs in the training regimen.


### Alternative Model 1

In [201]:
# Define the the number of inputs (features) to the model
number_input_features = len(X_train.iloc[0])

# Review the number of features
number_input_features

116

In [202]:
# Define the number of neurons in the output layer
number_output_neurons_A1 = 1

In [203]:
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1_A1 = 50

# Review the number of hidden nodes in the first layer
hidden_nodes_layer1_A1

50

In [204]:
# Create the Sequential model instance
nn_A1 = Sequential()

In [205]:
# First hidden layer
nn_A1.add(Dense(units=hidden_nodes_layer1_A1, activation='relu', input_dim=number_input_features))


# Output layer
nn_A1.add(Dense(units = number_output_neurons_A1, activation = 'linear'))


# Check the structure of the model
nn_A1.summary()

Model: "sequential_15"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_44 (Dense)            (None, 50)                5850      
                                                                 
 dense_45 (Dense)            (None, 1)                 51        
                                                                 
Total params: 5,901
Trainable params: 5,901
Non-trainable params: 0
_________________________________________________________________


In [206]:
# Compile the Sequential model
nn_A1.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])


In [207]:
# Fit the model using 50 epochs and the training data
fit_model_A1 = nn_A1.fit(X_train_scaled,y_train, epochs=50)


Epoch 1/50
804/804 [==============================] - 1s 577us/step - loss: 1.5516 - accuracy: 0.6742
Epoch 2/50
804/804 [==============================] - 0s 572us/step - loss: 1.9298 - accuracy: 0.6765
Epoch 3/50
804/804 [==============================] - 0s 573us/step - loss: 1.3281 - accuracy: 0.6841
Epoch 4/50
804/804 [==============================] - 0s 567us/step - loss: 1.6091 - accuracy: 0.6677
Epoch 5/50
804/804 [==============================] - 0s 566us/step - loss: 1.7953 - accuracy: 0.6686
Epoch 6/50
804/804 [==============================] - 0s 557us/step - loss: 1.6089 - accuracy: 0.6862
Epoch 7/50
804/804 [==============================] - 0s 566us/step - loss: 1.4378 - accuracy: 0.6855
Epoch 8/50
804/804 [==============================] - 0s 570us/step - loss: 1.2786 - accuracy: 0.7032
Epoch 9/50
804/804 [==============================] - 0s 561us/step - loss: 1.4710 - accuracy: 0.6941
Epoch 10/50
804/804 [==============================] - 0s 563us/step - loss: 1.528

#### Alternative Model 2

In [208]:
# Define the the number of inputs (features) to the model
number_input_features = len(X_train.iloc[0])

# Review the number of features
number_input_features

116

In [209]:
# Define the number of neurons in the output layer
number_output_neurons_A2 = 1

In [210]:
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1_A2 = 5

# Review the number of hidden nodes in the first layer
hidden_nodes_layer1_A2

5

In [211]:
# Create the Sequential model instance
nn_A2 = Sequential()

In [212]:
# First hidden layer
nn_A2.add(Dense(units=hidden_nodes_layer1_A2, input_dim=number_input_features, activation='relu'))

# Output layer
nn_A2.add(Dense(units=number_output_neurons_A2, activation='sigmoid'))

# Check the structure of the model
nn_A2.summary()

Model: "sequential_16"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_46 (Dense)            (None, 5)                 585       
                                                                 
 dense_47 (Dense)            (None, 1)                 6         
                                                                 
Total params: 591
Trainable params: 591
Non-trainable params: 0
_________________________________________________________________


In [213]:
# Compile the model
nn_A2.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])


In [214]:
# Fit the model
fit_model_2= nn_A2.fit(X_train_scaled,y_train, epochs=50)


Epoch 1/50
804/804 [==============================] - 1s 545us/step - loss: 0.6078 - accuracy: 0.7035
Epoch 2/50
804/804 [==============================] - 0s 544us/step - loss: 0.5665 - accuracy: 0.7264
Epoch 3/50
804/804 [==============================] - 0s 545us/step - loss: 0.5585 - accuracy: 0.7276
Epoch 4/50
804/804 [==============================] - 0s 548us/step - loss: 0.5550 - accuracy: 0.7288
Epoch 5/50
804/804 [==============================] - 0s 542us/step - loss: 0.5531 - accuracy: 0.7290
Epoch 6/50
804/804 [==============================] - 0s 545us/step - loss: 0.5521 - accuracy: 0.7299
Epoch 7/50
804/804 [==============================] - 0s 542us/step - loss: 0.5517 - accuracy: 0.7293
Epoch 8/50
804/804 [==============================] - 0s 543us/step - loss: 0.5511 - accuracy: 0.7290
Epoch 9/50
804/804 [==============================] - 0s 540us/step - loss: 0.5504 - accuracy: 0.7300
Epoch 10/50
804/804 [==============================] - 0s 548us/step - loss: 0.550

### Step 2: After finishing your models, display the accuracy scores achieved by each model, and compare the results.

In [218]:
print("Original Model Results")

# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy =nn.evaluate(X_test_scaled,y_test, verbose=2)

# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

Original Model Results
268/268 - 0s - loss: 0.5523 - accuracy: 0.7312 - 104ms/epoch - 388us/step
Loss: 0.5523036122322083, Accuracy: 0.731195330619812


In [219]:
print("Alternative Model 1 Results")

# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy =nn_A1.evaluate(X_test_scaled,y_test, verbose=2)

# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

Alternative Model 1 Results
268/268 - 0s - loss: 0.7178 - accuracy: 0.7273 - 159ms/epoch - 593us/step
Loss: 0.7178284525871277, Accuracy: 0.7273469567298889


In [220]:
print("Alternative Model 2 Results")

# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy = nn_A2.evaluate(X_test_scaled,y_test, verbose=2)

# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

Alternative Model 2 Results
268/268 - 0s - loss: 0.5600 - accuracy: 0.7317 - 155ms/epoch - 578us/step
Loss: 0.5599712133407593, Accuracy: 0.7316617965698242


## Model Attempt #3 had the best accuracy score with 73.17% accuracy. 

### Step 3: Save each of your alternative models as an HDF5 file.


In [221]:
# Set the file path for the first alternative model
file_path_1 = Path('Resources/Model_Alternative_1')

# Export your model to a HDF5 file
nn_A1.save(file_path_1)


INFO:tensorflow:Assets written to: Resources\Model_Alternative_1\assets


In [222]:
# Set the file path for the second alternative model
file_path_2 =Path('Resources/Model_alt_2')

# Export your model to a HDF5 file
nn_A2.save(file_path_2)

INFO:tensorflow:Assets written to: Resources\Model_alt_2\assets
